### PR TITLE
Grm integration

### PR DESCRIPTION
**Pull request by avidan-demisto**
## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/14044

## Description
Add Google Resource Manager integration to master

## Required version of Demisto
any

## Does it break backward compatibility?
   - No

## Must have
- [ ] Tests
- [ ] Documentation (with link to it)
- [ ] Code Review

## Dependencies
Mention the dependencies of the entity you changed as given from the precommit hooks in checkboxes, and tick after tested them.
- [x] Google Resource Manager

## Additional changes
Add GoogleResourceManager-Test playbook to TestPlaybooks directory

### CODE REVIEW

File: Integrations/integration-GoogleResourceManager.yml [hunk 1]
```diff
@@ -0,0 +1,1020 @@
+category: IT Services
+commonfields: {id: Google Resource Manager, version: -1}
+configuration:
+- defaultvalue: ''
+  display: The Project ID of the project resource in which you created the Service
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: Change all parameters display names to be like `Project ID` - no need to put all this description here

File: Integrations/integration-GoogleResourceManager.yml [hunk 2]
```diff
@@ -0,0 +1,1020 @@
...
+  type: 4
+- defaultvalue: ''
+  display: Value of the 'client_x509_cert_url' field in your JSON Service Account
+    Key File.
+  name: client_x509_cert_url
+  options: []
+  required: true
+  type: 0
+description: Google Cloud Platform Resource Manager
+detaileddescription: Use these detailed instructions in order to retrieve the API
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: Please edit this and put here instructions on how to configure the integration (i.e. where to get all the required parameters)

File: Integrations/integration-GoogleResourceManager.yml [hunk 3]
```diff
@@ -0,0 +1,1020 @@
+category: IT Services
+commonfields: {id: Google Resource Manager, version: -1}
+configuration:
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: Please add `proxy` and `unsecure` parameters

File: Integrations/integration-GoogleResourceManager.yml [hunk 4]
```diff
@@ -0,0 +1,1020 @@
...
+      defaultValue: ''
+      description: |-
+        The label values associated with this Project.
+        Label values must be between 0 and 63 characters long and must conform to the regular expression ([a-z]([-a-z0-9]*[a-z0-9])?)?. A label value can be empty.
+        No more than 256 labels can be associated with a given resource.
+      isArray: true
+      name: labels.values
+      predefined: []
+      required: false
+    description: Create Project Resource
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: please improve description

File: Integrations/integration-GoogleResourceManager.yml [hunk 5]
```diff
@@ -0,0 +1,1020 @@
...
+script:
+  commands:
+  - arguments:
+    - default: false
+      defaultValue: ''
+      description: |-
+        The unique, user-assigned ID of the Project. It must be 6 to 30 lowercase letters, digits, or hyphens. It must start with a letter. Trailing hyphens are prohibited.
+        Example: tokyo-rain-123
+      isArray: false
+      name: projectId
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: change to `project_id`

File: Integrations/integration-GoogleResourceManager.yml [hunk 6]
```diff
@@ -0,0 +1,1020 @@
...
+      predefined: ['']
+      required: true
+    - default: false
+      defaultValue: ''
+      description: |-
+        string
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: remove this `string`

File: Integrations/integration-GoogleResourceManager.yml [hunk 7]
```diff
@@ -0,0 +1,1020 @@
...
+        The id of the parent resource.
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: change to `ID` and elaborate on how to get that `ID`

File: Integrations/integration-GoogleResourceManager.yml [hunk 8]
```diff
@@ -0,0 +1,1020 @@
...
+      predefined: []
+      required: true
+    - default: false
+      defaultValue: ''
+      description: |-
+        The label keys associated with this Project.
+        Label keys must be between 1 and 63 characters long and must conform to the following regular expression: [a-z]([-a-z0-9]*[a-z0-9])?.
+        No more than 256 labels can be associated with a given resource.
+      isArray: true
+      name: labels.attrs
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: improve argument name, don't spare on characters and don't use dots, but underscores

File: Integrations/integration-GoogleResourceManager.yml [hunk 9]
```diff
@@ -0,0 +1,1020 @@
...
+        Label values must be between 0 and 63 characters long and must conform to the regular expression ([a-z]([-a-z0-9]*[a-z0-9])?)?. A label value can be empty.
+        No more than 256 labels can be associated with a given resource.
+      isArray: true
+      name: labels.values
+      predefined: []
+      required: false
+    description: Create Project Resource
+    execution: false
+    name: grm-create-project
+    outputs:
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: improve outputs descriptions

File: Integrations/integration-GoogleResourceManager.yml [hunk 10]
```diff
@@ -0,0 +1,1020 @@
...
+    - {contextPath: GRM.Project.Labels, description: Labels of the project, type: Unknown}
+    - {contextPath: GRM.Project.Parent.ID, description: id of the parent resource,
+      type: String}
+    - {contextPath: GRM.Project.Parent.Type, description: type of the parent resource,
+      type: String}
+  - arguments:
+    - default: false
+      defaultValue: ''
+      description: |-
+        An expression for filtering the results of the request. Filter rules are case insensitive. The fields eligible for filtering are:
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: Description is way too long..

File: Integrations/integration-GoogleResourceManager.yml [hunk 11]
```diff
@@ -0,0 +1,1020 @@
...
+      name: pageSize
+      predefined: []
+      required: false
+    - default: false
+      defaultValue: ''
+      description: |-
+        string
+        A pagination token returned from a previous call to `organizations.search` that indicates from where listing should continue.
+      isArray: false
+      name: pageToken
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: change to `page_token`

File: Integrations/integration-GoogleResourceManager.yml [hunk 12]
```diff
@@ -0,0 +1,1020 @@
...
+    - {contextPath: GRM.Project.Parent.Type, description: type of the parent resource,
+      type: String}
+  - arguments:
+    - default: false
+      defaultValue: ''
+      description: |-
+        number
+        The maximum number of Organizations to return in the response.
+      isArray: false
+      name: pageSize
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: change to `page_size`

File: Integrations/integration-GoogleResourceManager.yml [hunk 13]
```diff
@@ -0,0 +1,1020 @@
...
+      description: |-
+        string
+        A pagination token returned from a previous call to `organizations.search` that indicates from where listing should continue.
+      isArray: false
+      name: pageToken
+      predefined: []
+      required: false
+    - default: false
+      defaultValue: ''
+      description: "Used to filter the Organizations to return in the response. Filter\
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: Description too long

File: Integrations/integration-GoogleResourceManager.yml [hunk 14]
```diff
@@ -0,0 +1,1020 @@
...
+    # Params for assembling object of the Service Account Credentials File Contents
+    SERVICE_ACT_PROJECT_ID = demisto.params().get('project_id').encode('utf-8')
+    PRIVATE_KEY_ID = demisto.params().get('private_key_id').encode('utf-8')
+    PRIVATE_KEY = demisto.params().get('private_key').encode('utf-8')
+    CLIENT_EMAIL = demisto.params().get('client_email').encode('utf-8')
+    CLIENT_ID = demisto.params().get('client_id').encode('utf-8')
+    CLIENT_X509_CERT_URL = demisto.params().get('client_x509_cert_url').encode('utf-8')
+
+    AUTH_JSON = {
+        "type": "service_account",
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: `"` -> `'`

File: Integrations/integration-GoogleResourceManager.yml [hunk 15]
```diff
@@ -0,0 +1,1020 @@
...
+            body['name'] = project_body['name']
+        if 'labels.attrs' in keys and 'labels.values' in keys:
+            labels_attrs = argToList(project_body['labels.attrs'])
+            labels_values = argToList(project_body['labels.values'])
+            if len(labels_attrs) != len(labels_values):
+                err_msg = 'Label attrs array and Label values array do not match'\
+                    ' in length.\nThese arrays need to match in length because '\
+                    'each string in labels.attrs is assigned the value of the '\
+                    'string in the corresponding index in the labels.values array.'
+                raise IndexError(err_msg)
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: use `return_error` instead of `raise`

File: Integrations/integration-GoogleResourceManager.yml [hunk 16]
```diff
@@ -0,0 +1,1020 @@
...
+            else:
+                body['labels'] = {}
+                for lbl_tag, lbl_val in zip(labels_attrs, labels_values):
+                    body['labels'][lbl_tag] = lbl_val
+        return body
+
+
+    def poll_operation(operation):
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: we might want to use our `polling` mechanism here, how long does it take for this to run ?

File: Integrations/integration-GoogleResourceManager.yml [hunk 17]
```diff
@@ -0,0 +1,1020 @@
...
+            full_err_msg = "error code: {}\nerror message: {}\nerror details: {}".format(err_code, err_msg, err_details)
+            raise Exception(full_err_msg)
+
+
+    ''' MAIN FUNCTIONS '''
+
+
+    def test_module():
+        """If the list_projects_command executes successfully then the test completed and returns 'ok'"""
+        build_and_authorize()
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: this should be before the commands block, not in every function

File: Integrations/integration-GoogleResourceManager.yml [hunk 18]
```diff
@@ -0,0 +1,1020 @@
...
+            'Number': response.get('projectNumber'),
+            'State': response.get('lifecycleState'),
+            'CreateTime': response.get('createTime'),
+            'Parent': {
+                'ID': response.get('parent').get('id'),
+                'Type': response.get('parent').get('type')
+            },
+            'Labels': response.get('labels')
+        }
+        md = tableToMarkdown('Google Cloud Project Created', context)
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: add something like `successfully`

File: Integrations/integration-GoogleResourceManager.yml [hunk 19]
```diff
@@ -0,0 +1,1020 @@
...
+
+        demisto parameter: (string) projectId
+            The unique ID of the Project to restore
+
+        returns:
+            Deleted project resource object
+        """
+        project_id = demisto.args()['projectId']
+        response = delete_project(project_id)
+        if json.dumps(response) == '{}':
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: not very pythonic way for a condition, why not just `if response:` ?

File: Integrations/integration-GoogleResourceManager.yml [hunk 20]
```diff
@@ -0,0 +1,1020 @@
...
+                'Number': response.get('projectNumber'),
+                'State': response.get('lifecycleState'),
+                'CreateTime': response.get('createTime'),
+                'Parent': {
+                    'ID': response.get('parent').get('id'),
+                    'Type': response.get('parent').get('type')
+                },
+                'Labels': response.get('labels')
+            }
+            success_msg = 'The lifecycleState attribute of Project {} has been set to DELETE_REQUESTED'.format(project_id)
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: where can i see that state in the context?

File: Integrations/integration-GoogleResourceManager.yml [hunk 21]
```diff
@@ -0,0 +1,1020 @@
...
+        }
+        md = tableToMarkdown('Details of Google Cloud Project', context)
+        demisto.results({
+            'Type': entryTypes['note'],
+            'Contents': response,
+            'ContentsFormat': formats['json'],
+            'ReadableContentsFormat': formats['markdown'],
+            'HumanReadable': md,
+            'EntryContext': {
+                'GRM.Project(val.ID && val.ID == obj.ID)': context
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: `===`

File: Integrations/integration-GoogleResourceManager.yml [hunk 22]
```diff
@@ -0,0 +1,1020 @@
...
+                'State': project.get('lifecycleState'),
+                'CreateTime': project.get('createTime'),
+                'Parent': {
+                    'ID': project.get('parent').get('id'),
+                    'Type': project.get('parent').get('type')
+                },
+                'Labels': project.get('labels')
+            }
+            contexts.append(context)
+        md = tableToMarkdown("Projects", contexts)
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: improve title

File: Integrations/integration-GoogleResourceManager.yml [hunk 23]
```diff
@@ -0,0 +1,1020 @@
...
+
+
+    def update_project(project_id, project_body):
+        """Build service object and return the result of calling the API 'update' function for the projects resource."""
+        service = build_and_authorize()
+        operation = service.projects().update(projectId=project_id, body=project_body).execute()
+        return operation
+
+
+    # Does not work How I would think it would ---- NEEDS ATTENTION
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: remove comment

File: Integrations/integration-GoogleResourceManager.yml [hunk 24]
```diff
@@ -0,0 +1,1020 @@
...
+
+        demisto parameter: (list) labels.values
+            The label values to associate with this Project. The values are
+            assigned to their corresponding indexed key in labels.attrs
+
+        returns:
+            The updated Project resource object
+        """
+        project_id = demisto.args()['projectId']
+        project_body = demisto.args()
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: remove this

File: Integrations/integration-GoogleResourceManager.yml [hunk 25]
```diff
@@ -0,0 +1,1020 @@
...
+            'Name': response.get('name'),
+            'ID': response.get('projectId'),
+            'Number': response.get('projectNumber'),
+            'State': response.get('lifecycleState'),
+            'CreateTime': response.get('createTime'),
+            'Parent': {
+                'ID': response.get('parent').get('id'),
+                'Type': response.get('parent').get('type')
+            },
+            'Labels': response.get('labels')
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: please use singular values in context, should be `Label`

File: Integrations/integration-GoogleResourceManager.yml [hunk 26]
```diff
@@ -0,0 +1,1020 @@
...
+        contents = []
+        next_page = True
+        while next_page:
+            response = search_organizations(req_body)
+            contents.append(response)
+            for organization in response.get('organizations', []):
+                # Parse organization into context
+                context = {
+                    'Name': organization.get('name'),
+                    'DisplayName': organization.get('displayName'),
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: whats the difference between `Name` and `DisplayName` ?

File: Integrations/integration-GoogleResourceManager.yml [hunk 27]
```diff
@@ -0,0 +1,1020 @@
...
+            Used to filter the Organizations to return in the response
+
+        returns:
+            List of Organization resource objects that are visible to the user and satisfy the specified filter
+        """
+        req_body = demisto.args()
+        contexts = []
+        contents = []
+        next_page = True
+        while next_page:
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: please add some comments that explain what you're doing

File: Integrations/integration-GoogleResourceManager.yml [hunk 28]
```diff
@@ -0,0 +1,1020 @@
...
+    try:
+        if demisto.command() == 'test-module':
+            # This is the call made when pressing the integration test button.
+            test_module()
+        elif demisto.command() in commands.keys():
+            commands[demisto.command()]()
+
+    # Log exceptions
+    except Exception as e:
+        # Output HttpError errors from googleapiclient to the warroom nicely
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: remove all comments

File: Integrations/integration-GoogleResourceManager.yml [hunk 29]
```diff
@@ -0,0 +1,1020 @@
...
+        #     LOG.print_log()
+        #     return_error(e.message)
+        # # Otherwise output the exception itself to the warroom
+        # else:
+        #     LOG(str(e))
+        #     LOG.print_log()
+        #     return_error(str(e))
+        LOG(e.message)
+        LOG.print_log()
+        return_error(e.message)
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: `return_error` already calls `LOG`

File: Integrations/integration-GoogleResourceManager.yml [hunk 30]
```diff
@@ -0,0 +1,1020 @@
...
+            'HumanReadable': md,
+            'EntryContext': {
+                'GRM.Organization(val.Name && val.Name === obj.Name)': context
+            }
+        })
+
+
+    def create_folder(parent, name):
+        """Build service object and return the result of calling the API 'create' function for the folders resource."""
+        API_VERSION = 'v2'
```
Comment (Itay4) [inline Integrations/integration-GoogleResourceManager.yml]: Why is this here?

File: Integrations/integration-GoogleResourceManager.yml (hunk 19 above)
Comment (avidan-demisto) [reply Integrations/integration-GoogleResourceManager.yml]: upon successful deletion of a project, the api returns an empty json object
